docs/guides/upgrading-v3.md

```diff
@@ -0,0 +1,265 @@
+<!--
+title: Serverless Framework - Upgrading to v3
+menuText: Upgrading to v3
+menuOrder: 12
+layout: Doc
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/guides/upgrading-v3/)
+
+<!-- DOCS-SITE-LINK:END -->
+
+# Upgrading to Serverless Framework v3
+
+Serverless Framework v3 contains a few breaking changes that may impact some projects.
+
+This guide helps users upgrade from Serverless Framework v2 to v3.
+
+**Note:** Serverless Framework v3 isn't released yet. However, this guide helps to prepare to v3, or upgrade to the v3 beta version.
+
+## Am I impacted by breaking changes?
+
+Serverless Framework v2 signals any deprecated feature via a deprecation warning. The simplest way to upgrade to v3 is to:
+
+1. Upgrade Serverless Framework to the latest v2 version
+2. Run `serverless` commands in the project to see if there are any deprecation warnings
+
+Projects that do not have any deprecations can be immediately upgraded to v3. Projects that have deprecation warnings should first solve these deprecations, then upgrade to v3.
+
+## What about plugins?
+
+We have worked with the most popular plugins to make sure they support Serverless Framework v3. As long as a project has no deprecations, it can be safely upgraded to v3.
+
+That being said, some plugins need to be updated to be installable with v3. In most cases, it is a matter of allowing the plugin to be installed with Serverless Framework v3 in `package.json` (updating the `peerDependencies` requirement). Feel free to open an issue or pull request in the GitHub repository of the plugin.
+
+## Upgrading to v3
+
+First, upgrade to the latest v2 version and make sure that you do not see any deprecation warning.
+
+Then, to upgrade to Serverless Framework v3, run [the command that matches your installation method](../getting-started.md#upgrade):
+
+- MacOS/Linux: `serverless upgrade --major`
```

Comment:
I think this should clearly be indicated these method works only if user relies on _standalone_ installation.

I see a note to _run command that matches your installation method_, but I think many users may fast read. and just assume that they can magically upgrade on macOs with `serverless upgrade --major`, which will crash for majority of them (a small fraction of users, relies on standalone)

Reply:
This is based on this documentation: https://www.serverless.com/framework/docs/getting-started#upgrade

How can we improve both? What can I explain to users so that they know what to do?

Reply:
Documentation on_standalone_ installation will probably work better as distinct page or section, and not to be mixed with _npm_ installation, which doesn't need much documentation aside of basic install command (it's in scope of npm package manager to document its commands)

Reply:
Right but how can we explain that to users?

I.e. as a user, reading the upgrade guide, what should I do? How do I know what instruction to follow?

Reply:
> I.e. as a user, reading the upgrade guide, what should I do? How do I know what instructions to follow?

If user decided to install a standalone, then should follow its instructions. Upgrade instructions are dedicated for standalone, not for any other type of installations

If user is relying on npm, then doesn't need any specific instructions from us (aside from pointing that Framework is available in npm registry, and can be installed via `npm install -g serverless`)

Reply:
As a user, how can I know how my CLI is installed? (if I installed my CLI 6 months ago, I don't remember)

> If user is relying on npm, then doesn't need any specific instructions from us

They do, some people don't know how to use NPM by heart.

Reply:
> They do, some people don't know how to use NPM by heart.

Sure, we may add some more comprehensive section on how to deal with npm, but I would split those two sections, so _npm_ and _standalone_ are not mixed. Those are very different kind of installations. _standalone_ is where we setup _registry_ rules and provide an interface to install and update, so we need to document it.

While in case of _npm_, we simply publish to npm registry, and to consume that users may rely on different PM's. It can be npm, Yarn or PnP etc. Usually projects in their documentation just signal that package is available at npm registry, and having that users rely on instructions of PM of their choice on how to install and update.

Reply:
I think if user is not familiar with npm, then should just rely on _standalone_ installation.

IIRC the idea behind current instructions was to favor _standalone_ installation, and on a side, inform that _package_ is available in npm registry for those preferring that way (because it's what most developers are familiar with).

So in light of that, I think there's no real point in providing a comprehensive npm installation doc

Reply:
As discussed, I've pushed an update where NPM is the default method (I slightly updated the "Getting started" guide too).

I updated the upgrade guide too.

Let me know @medikoo & @pgrzesik